Fix checking ctx to carry correct modes

noti0na1 opened the pull request:
In `TreeChecker`, the context for checking is not initialized properly. `val checkingCtx` uses `setMode(Mode.ImplicitsEnabled)`, which removes other modes in the ctx. 

This causes `Mode.SafeNulls` removed from the context. So the checking for explicit nulls tests is always running with unsafe nulls enabled!!

This PR fixs the checking ctx and some other bugs exposed by the correct checking:

* Remove Imports during Erasure, so all other phases before it can be checked correctly
* Cast the last expression of a block to the expected type, 
  if unsafe nulls is enabled inside but not enabled outside 
  and the type does not conform the expected type without unsafe nulls. 
  See: `tests/explicit-nulls/pos/unsafe-block.scala`
* Use unsafe nulls when forwarding a member before erasure if the target is Java defined
* Cast `null` to `AnyRef` when generating `testNotNull`
* When creating `nameRef` in `SyntheticMembers.scala`, cast the result of `name` to `String` if `isJavaEnumValue` 

#15096 is blocked by this PR, #14946 will also wait for this

noti0na1 commented on compiler/src/dotty/tools/dotc/transform/Erasure.scala:
```diff
@@ -153,6 +153,7 @@ class Erasure extends Phase with DenotTransformer {
   override def checkPostCondition(tree: tpd.Tree)(using Context): Unit = {
     assertErased(tree)
     tree match {
+      case _: tpd.Import => assert(false, i"illegal tree: $tree")
```
Since DropImports is not checkable, the `checkPostCondition` will not be called

noti0na1 commented on compiler/src/dotty/tools/dotc/typer/Typer.scala:
```diff
@@ -1058,7 +1058,18 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
-    val expr1 = typedExpr(tree.expr, pt.dropIfProto)(using exprCtx)
+    var expr1 = typedExpr(tree.expr, pt.dropIfProto)(using exprCtx)
+
+    // If unsafe nulls is enabled inside a block but not enabled outside
```
Sorry I don't understand your question. This code tries to solve the issue at the boundary of an unsafe scope. 

```scala
// the expected type of the block is `T1`
val x: T1 = {
  ...
  import unsafeNulls
  ...
  // unsafeNulls is enabled at this point
  exp 
  // the type of exp is T2
  // T2 <:< T1 here, typed success
} // unsafeNulls is not enabled here,  T2 !<:< T1
```

olhotak replied:
After discussing offline, this is resolved: the issue is that the type of the block needs to be the same as the type of the last expression, so the change in type cannot happen at the boundary of the last expression and the block itself.

noti0na1 commented:
test performance please

olhotak commented on compiler/src/dotty/tools/dotc/Compiler.scala:
```diff
@@ -101,6 +101,7 @@ class Compiler {
          new TupleOptimizations,     // Optimize generic operations on tuples
          new LetOverApply,           // Lift blocks from receivers of applications
          new ArrayConstructors) ::   // Intercept creation of (non-generic) arrays and intrinsify.
+    List(new DropImports) ::         // Drop all (language-) imports
```
OK. I'd like @odersky 's opinion on adding this megaphase and possible alternatives. Otherwise LGTM.

noti0na1 commented:
test performance please

noti0na1 commented:
Hi @odersky , I removed the `DropImports` phase and drop imports in Erasure now. I also fixed two other bugs related to checking since last review.

This PR is ready for another review now. Thanks

olhotak reviewed (approved):
LGTM

odersky commented on compiler/src/dotty/tools/dotc/typer/Typer.scala:
```diff
@@ -1059,7 +1059,18 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
-    val expr1 = typedExpr(tree.expr, pt.dropIfProto)(using exprCtx)
+    var expr1 = typedExpr(tree.expr, pt.dropIfProto)(using exprCtx)
+
+    // If unsafe nulls is enabled inside a block but not enabled outside
```
The comment does not correspond to the code. `exprCtx` is the context inside the block valid for the final expression. `ctx` is the context outside the block. The cast is performed if unsafeNulls is _disabled_ inside the block but _enabled_ around it. Is this what's intended? Either the comment or the code has to be fixed.

noti0na1 replied:
I think the comment is correct: `!exprCtx.mode.is(Mode.SafeNulls)` indicates `SafeNulls` is disabled inside the block, which means unsafeNulls is enabled here.

olhotak replied:
The confusion is that the mode is `SafeNulls` which is a negation of `unsafeNulls`. I think the comment correctly corresponds to the code.

odersky replied:
Yes, my bad. I misread the code.

odersky reviewed (changes requested):
Otherwise LGTM
